source/module_hamilt_pw/hamilt_pwdft/fs_kin_tools.cpp

```diff
@@ -0,0 +1,160 @@
+{
+    this->device = base_device::get_device_type<Device>(this->ctx);
+    this->wfc_basis_ = wfc_basis_in;
+    const int npwk_max = this->wfc_basis_->npwk_max;
+    const int nks = this->wfc_basis_->nks;
+    const int npol = ucell_in.get_npol();
+
+    this->gk3_.resize(3 * npwk_max);
+    this->gk.resize(3);
+    for (int i = 0; i < 3; ++i)
+    {
+        this->gk[i] = &this->gk3_[i * npwk_max];
+    }
+    this->kfac.resize(npwk_max);
+    this->s_kin.resize(9, 0.0);
+
+    if (this->device == base_device::GpuDevice)
+    {
+        resmem_var_op()(this->ctx, d_gk, 3 * npwk_max);
+        resmem_var_op()(this->ctx, d_kfac, npwk_max);
+    }
+    else
+    {
+        d_gk = gk3_.data();
+        d_kfac = kfac.data();
+    }
+}
+
+template <typename FPTYPE, typename Device>
+FS_Kin_tools<FPTYPE, Device>::~FS_Kin_tools()
+{
+    if (this->device == base_device::GpuDevice)
+    {
+        delmem_var_op()(this->ctx, d_gk);
+        delmem_var_op()(this->ctx, d_kfac);
+    }
+}
+
+template <typename FPTYPE, typename Device>
+void FS_Kin_tools<FPTYPE, Device>::cal_gk(const int& ik)
+{
+    const double tpiba = ModuleBase::TWO_PI / this->ucell_.lat0;
+    const double twobysqrtpi = 2.0 / std::sqrt(ModuleBase::PI);
+    const int npw = wfc_basis_->npwk[ik];
+    const int npwk_max = wfc_basis_->npwk_max;
+    for (int i = 0; i < npw; ++i)
+    {
+        gk[0][i] = wfc_basis_->getgpluskcar(ik, i)[0] * tpiba;
+        gk[1][i] = wfc_basis_->getgpluskcar(ik, i)[1] * tpiba;
+        gk[2][i] = wfc_basis_->getgpluskcar(ik, i)[2] * tpiba;
+        if (wfc_basis_->erf_height > 0)
+        {
+            double gk2 = gk[0][i] * gk[0][i] + gk[1][i] * gk[1][i] + gk[2][i] * gk[2][i];
+            double arg = (gk2 - wfc_basis_->erf_ecut) / wfc_basis_->erf_sigma;
+            kfac[i] = 1.0 + wfc_basis_->erf_height / wfc_basis_->erf_sigma * twobysqrtpi * std::exp(-arg * arg);
+        }
+        else
+        {
+            kfac[i] = 1.0;
+        }
+    }
+    if (this->device == base_device::GpuDevice)
+    {
+        syncmem_var_h2d_op()(this->ctx, this->cpu_ctx, d_gk, gk[0], 3 * npwk_max);
+        syncmem_var_h2d_op()(this->ctx, this->cpu_ctx, d_kfac, kfac.data(), npwk_max);
+    }
+}
+
+template <typename FPTYPE, typename Device>
+void FS_Kin_tools<FPTYPE, Device>::cal_stress_kin(const int& ik,
+                                                  const int& npm,
+                                                  const bool& occ,
+                                                  const std::complex<FPTYPE>* psi)
+{
+    if (npm == 0)
+    {
+        return;
+    }
+    const int npw = wfc_basis_->npwk[ik];
+    const int npwk_max = wfc_basis_->npwk_max;
+    const int npol = this->ucell_.get_npol();
+    for (int ib = 0; ib < npm; ib++)
+    {
+        const std::complex<FPTYPE>* ppsi = psi + ib * npwk_max * npol;
+        const std::complex<FPTYPE>* ppsi2 = ppsi + npwk_max;
+        FPTYPE fac;
```

Comment:
is it possible to initialize this variable?

Reply:
OK